src/borg/_hashindex.c

```diff
@@ -376,29 +379,69 @@ hashindex_get(HashIndex *index, const void *key)
     return BUCKET_ADDR(index, idx) + index->key_size;
 }
 
+inline int
+distance(HashIndex *index, int current_idx, int ideal_idx)
+{
+    /* If the current index is smaller than the ideal index we've wrapped
+       around the end of the bucket array and need to compensate for that. */
+    return current_idx - ideal_idx + ( (current_idx < ideal_idx) ? index->num_buckets : 0 );
+}
+
+inline void
+memswap(void *a, void *b, void *tmp, size_t entry_size) {
+    memcpy(tmp, a, entry_size);
+    memcpy(a, b, entry_size);
+    memcpy(b, tmp, entry_size);
+}
+
 static int
 hashindex_set(HashIndex *index, const void *key, const void *value)
 {
     int idx = hashindex_lookup(index, key);
-    uint8_t *ptr;
+    uint8_t *bucket_ptr;
+    int offset = 0;
+    int other_offset;
+    int entry_size = (index->key_size + index->value_size);
+    static void *entry_to_insert = NULL;
+    static void *tmp_entry = NULL;
+    if (entry_to_insert == NULL) {
+        entry_to_insert = malloc(entry_size * 2);
+        tmp_entry = entry_to_insert + entry_size;
+    }
```

Comment:
hmm, is this code correct if used for different kinds of HashIndexes (different entry sizes)?
if it is called first needing entry_size X, it will allocate a 2X sized buffer, but if it is afterwards called needing entry_size Y, it will reuse that buffer...

Also, it only has one buffer, so multithreaded usage of this might lead to problems also (maybe not currently, but in future?).